pkg/compute/logstream/execution_log_writer.go

```
@@ -0,0 +1,44 @@
+package logstream
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/bacalhau-project/bacalhau/pkg/compute"
+	"github.com/bacalhau-project/bacalhau/pkg/util"
+	"github.com/bacalhau-project/bacalhau/pkg/util/closer"
+)
+
+type ExecutionLogWriterParams struct {
+	logsDir string
+}
+
+type ExecutionLogWriter struct {
+	params *ExecutionLogWriterParams
+}
+
+func NewExecutionLogWriter(logsDir string) (*ExecutionLogWriter, error) {
+	return &ExecutionLogWriter{&ExecutionLogWriterParams{logsDir}}, nil
+}
+
+func (w *ExecutionLogWriter) StartWriting(src io.Reader) chan util.Result[int64] {
+	resultCh := make(chan util.Result[int64])
+	go func() {
+		defer close(resultCh)
+		fileWriter, err := os.Create(filepath.Join(w.params.logsDir, compute.ExecutionLogFileName))
+		if err != nil {
+			resultCh <- util.NewResult[int64](0, err)
+			return
+		}
+		defer closer.CloseWithLogOnError("executionLogWriter", fileWriter)
+
+		copiedBytes, err := StdCopyWithEndFrame(fileWriter, src)
+		// Write the result to the channel, but don't block if the channel is full or nothing is reading.
+		select {
+		case resultCh <- util.NewResult(copiedBytes, err):
+		default:
+		}
```

use buffered channel to avoid blocking instead of select and default, as this will drop any error 
```
ch := make(chan error, 1) // buffered with 1 slot

ch <- errors.New(...)    // OK
ch <- errors.New(...)  // blocks, because buffer is full, but you don't write two errors
```

---

* `ExecutionLogReader` no longer exposes a channel, it uses a context instead to monitor when it needs to stop reading.
* As a side note I wanted to discuss this buffered channel vs non-blocking channel operations. As you know I'm pretty new to the language so don't know yet what's considered more idiomatic. I agree that non-blocking reading looks a bit hacky compared to buffered channel but it seems to be a fairly recognized pattern ([1](https://gobyexample.com/non-blocking-channel-operations), [2](https://stackoverflow.com/questions/48953236/how-to-implement-non-blocking-write-to-an-unbuffered-channel), [3](https://www.geeksforgeeks.org/go-non-blocking-channel-operations/)). Buffered channel will also block if something reads/writes more times than the buffer size. The non-blocking pattern does not have that problem. I haven't formed my own opinion yet so I'm curious to hear what you think.

---

oh yeah, both are acceptable and valid patterns, and it depends on the context. The non-blocking means it is best effort and if there was no reader blocking on channel at that time, then it will drop the error.

What this means in your case, if an error was thrown immediately in your LogReader for example and before LiveStreamer or any other component that will be reading from the channel was created, then the error will be dropped, and then when trying to read from the channel you wouldn't know what happened.

The buffered channel does have the limitation that it will block if you have more than one error to write, but again it depends on your use case. Here for example you always report a single async error. The buffered channel guarantees that an error is reported regardless if there was an active and blocking consumer or not